sympy/sets/fancysets.py

```diff
@@ -532,14 +536,17 @@ def __len__(self):
         if not self:
             return 0
         dif = self.stop - self.start
-        if dif.is_infinite:
+        if dif.is_infinite or Tuple(dif, self.step).has(Symbol):
             raise ValueError(
-                "Use .size to get the length of an infinite Range")
+                "Use .size to get the length of an infinite and symbolic Range.")
         return abs(dif//self.step)
 
     @property
     def size(self):
         try:
+            dif = self.stop - self.start
```

Comment:
This doesn't always work:
```python
>>> eq = abs((stop - start)//step)
>>> for i,j,k in ((1, 5, 2), (1, 6, 2), (1, 5, -2), (6, 1, -2), (6, 0, -2), (6, 0, 2)):
...     print (i,j,k,Eq(eq.subs(dict(zip((start, stop, step), (i,j,k)))), len(range(i,j,k))))
(1, 5, 2, True)
(1, 6, 2, False)
(1, 5, -2, False)
(6, 1, -2, False)
(6, 0, -2, True)
(6, 0, 2, False)
```

Reply:
This means that the start and stop values are not being calculated correctly at instantiation to be symbolically true. When they are then this test should pass:
```python
r = Range(a,b,c).size
for i,j,k in (
  (1, 5, 2), (1, 6, 2), (1, 5, -2), (6, 1, -2), (6, 0, -2), (6, 0, 2)):
  print (i,j,k,Eq(r.subs(dict(zip((a,b,c), (i,j,k)))), len(range(i,j,k))))
```

Reply:
here is what I get
```python
        # expand range
        slc = slice(*args)

        if slc.step == 0:
            raise ValueError("step cannot be 0")

        start, stop, step = slc.start or 0, slc.stop, slc.step or 1

        start, stop, step = map(_sympify, [start, stop, step])

        if step.is_infinite:
            raise ValueError(filldedent('''
    Range cannot have an infinite step size.'''))

        # make literal
        t = [start, stop, step]
        for i, ti in enumerate(t):
          if ti.is_zero:
            t[i] = S.Zero
          elif ti.is_infinite:
            if ti.is_positive:
              t[i] = oo
            elif it.is_negative:
              t[i] = -oo
            else:
              raise ValueError('infinite values must be signed')
        del t

        if all(i.is_infinite for i in  (start, stop)):
            if start == stop:
                # canonical null handled below
                start = stop = S.One
            else:
                raise ValueError(filldedent('''
    Either the start or end value of the Range must be finite.'''))

        if start.is_infinite:
            if step.is_negative and start is -oo:
              start = stop = S.One
            elif step.is_positive and start is oo:
              start = stop = S.One
            else:
              end = stop
        elif stop.is_infinite:
            if step.is_negative and stop is oo:
              start = stop = S.One
            elif step.is_positive and stop is -oo:
              start = stop = S.One
        if not any(i.is_infinite for i in (start, stop)):
            t = step, stop, start
            if not all(i.is_Symbol or i.is_Integer for i in t):
                raise ValueError(filldedent('''
Only Symbol and Integer are permitted in Range.'''))
            from sympy import Piecewise
            n = Piecewise((ceiling((stop - start)/step), Gt((stop-start)*step,0)),(0, True))
            if (n <= 0) == True:
                # null Range
                start = end = S.Zero
                step = S.One
            else:
                end = start + n*step
        return Basic.__new__(cls, start, end, step)

    start = property(lambda self: self.args[0])
    stop = property(lambda self: self.args[1])
    step = property(lambda self: self.args[2])

    def _contains(self, other):
        if not self:
            return S.false
        if other.is_infinite:
            return S.false
        if not other.is_integer:
            return S.false
        ref = self.stop if self.start.is_infinite else self.start
        return And(Eq((ref - other) % self.step, 0), Ge(other, self.inf), Le(other, self.sup))

    def __iter__(self):
        if not all(i.is_Integer for i in self.args):
            raise ValueError("Cannot iterate over infinite or symbolic Range.")
        elif self:
            i = self.start
            step = self.step

            while True:
                if (step > 0 and not (self.start <= i < self.stop)) or \
                   (step < 0 and not (self.stop < i <= self.start)):
                    break
                yield i
                i += step

    def __len__(self):
        if not self:
            return 0
        dif = self.stop - self.start
        if dif.is_infinite or Tuple(dif, self.step).has(Symbol):
            raise ValueError(
                "Use .size to get the length of an infinite and symbolic Range.")
        return dif//self.step

    @property
    def size(self):
        try:
            return (self.stop - self.start)//self.step
        except ValueError:
            return S.Infinity
```

Reply:
@smichr Can you please send me the `diff`? I am working on it and from what I have done is breaking many tests. Your diff can speed up the work.

Reply:
I am working on the code your provided. I will update you by tomorrow on this.

Reply:
> send me the `diff`

no diff - I just copied and pasted the code

Reply:
No problem. I have used the code. Some bugs are there. I will fix them and push the changes asap.